types/model/name.go

```diff
@@ -0,0 +1,630 @@
+//
+//	"example.com/mistral:7b+"
+//	"example.com/mistral:7b+Q4_0+"
+//	"x/y/z/z:8n+I"
+//	""
+//
+// It returns the zero value if any part is invalid.
+//
+// As a rule of thumb, an valid name is one that can be round-tripped with
+// the [Name.String] method. That means ("x+") is invalid because
+// [Name.String] will not print a "+" if the build is empty.
+func ParseName(s string) Name {
+	var r Name
+
+	// Once rangefunc is enabled, we can replace this with:
+	//
+	// for kind, part := range Parts(s) { ... }
+	parts(s)(func(kind PartKind, part string) bool {
+		if kind == PartInvalid {
+			r = Name{}
+			return false
+		}
+		if kind == PartDigest && !ParseDigest(part).IsValid() {
+			r = Name{}
+			return false
+		}
+		r.parts[kind] = part
+		return true
+	})
+	if r.IsValid() || r.IsResolved() {
+		return r
+	}
+	return Name{}
+}
+
+func MustParseName(s string) Name {
+	r := ParseName(s)
+	if !r.IsValid() {
+		panic("model.MustParseName: invalid name: " + s)
+	}
+	return r
+}
+
+// Fill fills in the missing parts of dst with the parts of src.
+//
+// The returned Name will only be valid if dst is valid.
+func Fill(dst, src Name) Name {
+	var r Name
+	for i := range r.parts {
+		r.parts[i] = cmp.Or(dst.parts[i], src.parts[i])
+	}
+	return r
+}
+
+// WithBuild returns a copy of r with the build set to the given string.
+func (r Name) WithBuild(build string) Name {
+	r.parts[PartBuild] = build
+	return r
+}
+
+func (r Name) WithDigest(digest Digest) Name {
+	r.parts[PartDigest] = digest.String()
+	return r
+}
+
+var mapHashSeed = maphash.MakeSeed()
+
+// MapHash returns a case insensitive hash for use in maps and equality
+// checks. For a convenient way to compare names, use [Name.EqualFold].
+//
+//nolint:errcheck
+func (r Name) MapHash() uint64 {
+	// correctly hash the parts with case insensitive comparison
+	var h maphash.Hash
+	h.SetSeed(mapHashSeed)
+	for _, part := range r.Parts() {
+		// downcase the part for hashing
+		for i := range part {
+			c := part[i]
+			if c >= 'A' && c <= 'Z' {
+				c = c - 'A' + 'a'
+			}
+			h.WriteByte(c)
+		}
+	}
+	return h.Sum64()
+}
+
+func (r Name) slice(from, to PartKind) Name {
+	var v Name
+	copy(v.parts[from:to+1], r.parts[from:to+1])
+	return v
+}
+
+// DisplayModel returns a display string composed of the model only.
+func (r Name) DisplayModel() string {
+	return r.parts[PartModel]
+}
+
+// DisplayFullest returns the fullest possible display string in form:
+//
+//	<host>/<namespace>/<model>:<tag>
+//
+// If any part is missing, it is omitted from the display string.
+//
+// It does not include the build part. For the fullest possible display
+// string with the build, use [Name.String].
+func (r Name) DisplayFullest() string {
+	return r.slice(PartHost, PartTag).String()
+}
+
+// DisplayShort returns the fullest possible display string in form:
+//
+//	<model>:<tag>
+//
+// If any part is missing, it is omitted from the display string.
+func (r Name) DisplayShort() string {
```

Comment:
short name should omit host and namespace iff they're default

Reply:
This does not show the host or namespace.

Reply:
What I mean is a short name omits host and namespace iff they're default values. If they're not default, a short name must include those values. E.g. these are all short names:

```
mistral:latest
llama2:7b-chat-q4_0
mike/discollama:latest
localhost:5000/mike/discollama:latest
```

This is to avoid collision between models in different namespaces or hosts, e.g.

```
[registry.ollama.ai/library/]mistral:latest
[registry.ollama.ai/]mike/mistral:latest
localhost:5000/library/mistral:latest
localhost:5000/mike/mistral:latest
```